File: commands/Fun/urban.js
const { Command, RichDisplay } = require("klasa");
const { MessageEmbed } = require("discord.js");
const superagent = require("superagent");

class Urban extends Command {
  constructor(...args) {
    super(...args, {
      description: "Search a term in urban dictionary!",
      cooldown: 5,
      runIn: ["text"],
      aliases: ["ud", "urbandictionary", "urbandict"],
      usage: "<term:string>",
      requiredPermissions: ["MANAGE_MESSAGES", "EMBED_LINKS", "ADD_REACTIONS"]
    });
    this.handlers = new Map();
  }
  
  async run(msg, [term]) {
    if(!msg.channel.nsfw) throw "You can use urban in nsfw channels only! sorry but it has inappropriate words, which could be against Discord ToS (Terms of Services).";
    
    const res = await superagent.get(`https://api.urbandictionary.com/v0/define?term=${term}`).catch(() => null);
    if(!res) throw "Something went wrong with urban, try again later.";
    const previous = this.handlers.get(msg.author.id);
    if(previous) previous.stop();
    const display = new RichDisplay(
      new MessageEmbed()
        .setTitle("Urban Dictionary")
        .setColor(0xff0000)
    );
    
    for(const data of res.body.list) {
      const format = `${this.format(data.definition)}\n\n*${this.format(data.example)}*`.length > 1980 ? false : true;
      const definition = format ? this.format(data.definition) : data.definition;
      const example = format ? this.format(data.example) : data.example;
      display.addPage((em) => em.setDescription(`${definition}\n\n*${example}*\n\n**Votes**\n:thumbsup: ${data.thumbs_up} :thumbsdown: ${data.thumbs_down}\n\nDefinition written by **${data.author}**`)); 
    }
    
    const handler = await display.run(await msg.send("Loading urban..."), {
      filter: (reaction, user) => user.id === msg.author.id,
      time: 1000 * 60 * 3
    });
    handler.on("end", () => this.handlers.delete(msg.author.id));
    this.handlers.set(msg.author.id, handler);
    return handler;
  }

  format(str) {
    // https://stackoverflow.com/questions/52374809/javascript-regular-expression-to-catch-boxes
    return str.replace(/\[([^\][]+)\]/g, (x, y) => `${x}(https://www.urbandictionary.com/define.php?term=${y.replace(/\s+/g, "+")})`);
  }
}

module.exports = Urban;

File: commands/fun/urban.go
package fun

import (
  "github.com/sapphire-cord/sapphire"
  "github.com/pollen5/urban-go"
  "github.com/pollen5/taiga/utils"
  "fmt"
)

// Search up urban dictionary for a term.
// Usage: <term:string...>
// Aliases: ud, urbandict, urbandictionary
func Urban(ctx *sapphire.CommandContext) {
  if !ctx.Channel.NSFW {
    ctx.Reply("You can use urban in NSFW channels only! Sorry but it has inappropriate words which could be against Discord's ToS (Terms of Services).")
    return
  }

  res, err := urban.Define(ctx.JoinedArgs())

  if err != nil {
    ctx.Error(err)
    return
  }

  p := sapphire.NewPaginatorForContext(ctx)

  p.SetTemplate(func() *sapphire.Embed {
    return sapphire.NewEmbed().
      SetTitle("Urban Dictionary").
      SetColor(0xDFAC7C)
  })

  for _, entry := range res {
    format := fmt.Sprintf("%s\n\n*%s*",
      urban.Markdown(entry.Definition),
      urban.Markdown(entry.Example))

    if len(format) > 2000 {
      format = fmt.Sprintf("%s\n\n*%s*", entry.Definition, entry.Example)
    }

    format = utils.Trim(format, 2000)

    p.AddPageString(fmt.Sprintf("%s\n\n**Votes** :thumbsup: %d :thumbsdown: %d\n\nDefinition written by **%s**",
      format,
      entry.ThumbsUp,
      entry.ThumbsDown,
      entry.Author,
    ))
  }

  p.Run()
}
